homeassistant/components/media_player/snapcast.py

```diff
@@ -0,0 +1,88 @@
+"""
+homeassistant.components.media_player.snapcast
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Provides functionality to interact with Snapcast clients.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/media_player.snapcast/
+"""
+
+import logging
+
+from homeassistant.const import (
+    STATE_ON, STATE_OFF)
+
+from homeassistant.components.media_player import (
+    MediaPlayerDevice,
+    SUPPORT_VOLUME_SET, SUPPORT_VOLUME_MUTE)
+
+SUPPORT_SNAPCAST = SUPPORT_VOLUME_SET | SUPPORT_VOLUME_MUTE
+DOMAIN = 'snapcast'
+REQUIREMENTS = ['snapcast==1.0.0']
+_LOGGER = logging.getLogger(__name__)
+
+
+# pylint: disable=unused-argument
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """ Sets up the Snapcast platform. """
+    import snapcast
+    import socket
+    host = config.get('host')
+    port = config.get('port', snapcast.CONTROL_PORT)
+    if not host:
+        _LOGGER.error('No snapserver host specified')
+        return
+    try:
+        server = snapcast.Snapserver(host, port)
+        add_devices([SnapcastDevice(client) for client in server.clients])
+    except socket.gaierror:
+        _LOGGER.error('Could not connect to Snapcast server at %s:%d',
+                      host, port)
+
+
+class SnapcastDevice(MediaPlayerDevice):
+    """ Represents a Snapcast client device. """
+
+    def __init__(self, client):
+        self._client = client
+
+    @property
+    def name(self):
+        """ Device name. """
+        return self._client.identifier
+
+    @property
+    def volume_level(self):
+        """ Volume level. """
+        return self._client.volume / 100
+
+    @property
+    def is_volume_muted(self):
+        """ Volume muted. """
+        return self._client.muted
+
+    @property
+    def supported_media_commands(self):
+        """ Flags of media commands that are supported. """
+        return SUPPORT_SNAPCAST
+
+    @property
+    def state(self):
+        """ State of the player. """
+        if self._client.connected:
+            return STATE_ON
+        return STATE_OFF
+
+    def mute_volume(self, mute):
+        """ Mute status. """
+        self._client.muted = mute
+
+    def set_volume_level(self, volume):
+        """ Volume level. """
+        self._client.volume = int(volume * 100)
```

Comment:
You may want to consider rounding here. Right now, if HA tries to set the volume to 77.8%, this will set the volume to 77% instead of 78%. Not particularly important and very nit-picky, but it is something to consider.